### ISSUE TITLE
Don't include "guest" in the Roles inputfield

### ISSUE THREAD

**Issue by Toutouwai** [processwire/processwire-requests#200]
### Short description of the enhancement

The Roles inputfield includes the guest role, but this role is always greyed out because every user is automatically a guest.

![2018-06-04_131704](https://user-images.githubusercontent.com/1538852/40893982-469fd458-67fa-11e8-8dc1-fdb58e156f75.png)

Although I can see why "guest" is included in the inputfield, that option serves no useful purpose. If it was removed it would have no negative effect - PW knows that every user must have the guest role so even if that checkbox is removed from the inputfield and the form submitted the guest role is still retained as expected.

But if the guest option were removed it would have the benefit that it would be much easier to make Roles a required field. When you have non-superusers that are tasked with creating users and assigning roles to them it is useful to have Roles be a required field so the user sees an error notification if they do not select one or more roles. I can't think of any scenario where it would be correct for a user to be created and given only the guest role.

Currently making Roles required doesn't work as expected because the selected guest role counts as the inputfield being filled out.

**Comment by adrianbj** [comment]
I completely agree.

For reference, here is a related discussion: https://github.com/ryancramerdesign/ProcessWire/issues/588

**Comment by Toutouwai** [comment]
Thanks Adrian, that's an interesting discussion. 

I'd rather not see "guest" listed in the Roles column of ProcessUser either, again because it doesn't serve a useful purpose. If you're a developer who is familiar with the workings of PW then you sort of know what guest means, but in that case you don't really need to see it listed. If you're a user (non-superuser) then you probably have no idea what guest means, so in that case it just adds confusion.

I'm sure there must be some purpose to the guest role but currently I can't see what that purpose is. Is it possible for a user to not be a guest? If not then why does the role exist? How would things be different if that role was not tracked or used anywhere in PW? Hopefully @ryancramerdesign can explain.

**Comment by teppokoivula** [comment]
> I'm sure there must be some purpose to the guest role but currently I can't see what that purpose is. Is it possible for a user to not be a guest? If not then why does the role exist?

In the issue linked above Ryan explains that the guest role is mostly a remnant from earlier times, but that being said, I can see at least one obvious use case for it: whenever you set permissions for users, guest is used to signify "all users".

Sure, we could just say "all users", or "everyone", or whatever – but wouldn't that just mean replacing guest with something else? How is that an improvement? :)

As for the issue at hand, I'm a bit conflicted:

- I've had some content editors ask what that role is and why it's there. It's easy to explain ("that's a default role that everyone gets"), but it _does_ still need explaining, and in that way can be seen as a source of confusion.

- On the other hand I don't personally mind seeing the guest role: as long as it exists, i.e. it's a real role, I would rather see it as a (non-editable) role in the GUI than not see it at all.

Every user having guest role is clear once you've read the docs or browsed enough code, but generally speaking "magic stuff", such as roles that are there but won't be displayed anywhere, tend to confuse (particularly) new users.

**Comment by Toutouwai** [comment]
I agree with Ryan's comment in the other thread and would like to see it implemented:

> Guest role probably shouldn't be displayed anywhere, except in the Roles editor and in the Template Access tab. Anywhere else, it means absolutely nothing and is a waste of screen real estate ... we should move towards elimination of guest in 2.6.

Seeing guest in the admin is fine for the developer/superuser, and the current users/roles interface serves those individuals quite well: granular permissions, cumulative roles, etc.

On the other hand, I think clients are not well served by the current users/roles interface. When you have the situation that the client needs to create and manage users (which is not unusual I think) then you want things to be as simple as possible with nothing on screen that isn't relevant or useful.

Currently I have following needs which I am working through but are proving more difficult to implement than I think they should be:

* Roles inputfield is required
* Only roles that the client may actually assign appear in the roles inputfield (i.e. no guest, no "higher" roles)
* Client selects one role **or** another - PW should make this scenario easier to implement because I think it's quite a common one. Cumulative roles are great for the superuser but more advanced than most clients need or should have to think about.
* Guest does not appear in Roles column of ProcessUser. This one is proving a real headache and I think I may have to resort to an additional custom roles field just to show in ProcessUser. There seems to be a lister bug that I'll file a separate issue for. 

-----

Update: I ended up solving the needs above by creating a new pseudo-roles field on the user template and then setting the actual roles for the user in a saveReady hook. This was a lot simpler than my previous approach of modifying the Roles field with hooks. I still think it would be good for the PW core to provide an option for using the Roles inputfield as a single page field rather than a multiple page field.
